### ISSUE TITLE
Bottleneck in mel / segment calculation?

### ISSUE THREAD

**Issue by sbuser** [Blair-Johnson/batch-whisper#6]
I gave this a try today on a large number of audio files with batch sizes of 16 and then 32.

If I watch nvidia-smi dmon while the batch processing runs, I see very low GPU memory/core usage for large periods of time, then a spike in both, followed by a batch of text. 

I speculated that there was a bottleneck in the decoding of the audio files into mels - but it looks to me like you do that at the start. I'm not familiar enough with what is calculated on and off the GPU and when here. Is it possible there's a lot of work involved decoding, eg, an MP3 that can be threaded that I'm missing somewhere? GPU utilization is only near maximum for maybe like 5% of the time the program is running which isn't my typical experience with ML resource utilization?

**Comment by Blair-Johnson** [comment]
You're absolutely right that there's an audio decoding bottleneck. Currently the audio files are decoded serially at the beginning of a batch with ffmpeg. This could and should be parallelized, as it can take a long time for long (or many) audio files. 

The GPU usage is actually quite descent once the audio files are decoded and the system begins transcription, but that part of the pipeline can't start until the tensor arrays with the decoded audio are prepared. Other parts of the pipeline definitely warrant investigation, but I think this explains what you're observing. 

Would you be interested in submitting a PR? The fix should be as simple as replacing the serial decoding of a batch of audio file-paths here with a multiprocessing implementation: https://github.com/Blair-Johnson/batch-whisper/blob/76b3f818fb94a7c18c37d2ffecf71d7c69182749/whisper/transcribe.py#L335

**Comment by sbuser** [comment]
I can probably do that for the calculation of mels, but that isn't where I'm seeing the bottleneck as I run transcribe - which was confusing to me. If that's the only place audio decoding is taking place, then something else is CPU bound that I'm not understanding.

I didn't do a good job of explaining. When run in verbose mode which a batch size of 32, text is output let's say every 30 seconds or so. nvidia-smi dmon during that 30 second window looks something like below. Then there will be a spike of GPU usage (the 99% here and 37% memory), and then some text will pop out.

`   
    0    60    58     -    19     0     0     0  5005  1581
    1    10    36     -     0     0     0     0   405     4
    0    61    58     -    35     7     0     0  5005  1441
    1    10    36     -     0     0     0     0   405     2
    0    60    58     -    14     0     0     0  5005  1378
    1    10    36     -     0     0     0     0   405     3
    0    61    58     -    16     0     0     0  5005  1378
    1    11    36     -     0     0     0     0   405     2
    0   226    66     -    99    37     0     0  5005  1403
    1    10    35     -     0     0     0     0   405     2
    0    72    60     -    28     6     0     0  5005  1747
    1    10    36     -     0     0     0     0   405     2
    0    72    59     -    15     0     0     0  5005  1769
    1    10    35     -     0     0     0     0   405     4
    0    72    60     -    19     0     0     0  5005  1771
`

Meanwhile I a show a single 5ghz CPU core pegged. So I'm trying to figure out what is CPU bound for the 30 seconds that prevents the GPU from working at 99% core usage and 37% memory more of the time.

If it was just the initial calculation of mels, wouldn't I only see that spike for new files? This pattern happens in the middle of a group of 32 files - then, yes, there is a slowdown while mels are calculated initially - but these are long periods of time here for decoding.

I don't really understand the model and what's being calculated where but thought you might and could point me in the right direction?

**Comment by sbuser** [comment]
Here's one slow portion, I think:

https://github.com/Blair-Johnson/batch-whisper/blob/76b3f818fb94a7c18c37d2ffecf71d7c69182749/whisper/decoding.py#L704-L707

As it's looping through samples, this is particularly slow but doesn't seem to be part of the model inference or taking place on a GPU that I can tell? I'm not familiar enough with the where the matrix math gets done here.

I'll keep investigating, but this appears to be very slow relative to the surrounding work.

**Comment by Blair-Johnson** [comment]
Good catch, I think that I found the issue. On this line... https://github.com/Blair-Johnson/batch-whisper/blob/76b3f818fb94a7c18c37d2ffecf71d7c69182749/whisper/decoding.py#L493
...the logit filter lists are are all initialized with the same list object. As a result, each new batch of tokens for each audio file was being filtered by **all** logit filters introducing redundancy and potential accuracy issues. I'll post a PR with that fix and the ffmpeg multiprocessing modification. Let me know if that improves performance on your end.

**Comment by sbuser** [comment]
Ah, That probably also explains why I have a list of `no_speech_prob` in the output as opposed to just the single probability I expected and get with a normal whisper run.

**Comment by sbuser** [comment]
I think the large disparity between our resource use was because I was using model=base and you were using large. When I switch to large, I get much higher GPU utilization which makes sense. ApplyTimestampRules is definitely the bottleneck with those lower parameter models.

Your fix here did raise GPU utilization but in the results I still have an array of no_speech_prob corresponding to the current batch size (which changes as various audio files fall off from length differences). 

I'll open another issue if I pin this down but there's also something going on with multiple model.decode() runs for each decode_with_fallback() toward the end of the batch - probably something to do with shifting file sizes. I can't tell yet if there's missing transcriptions and it's just falling back or what.

**Comment by Blair-Johnson** [comment]
I updated the PR to address the `no_speech_prob` issue, let me know if you have any issues. As for the multiple `model.decode()` executions, could you verify whether they are being called because of temperature fallback? The loop checks to see if the compression ratio of generated text is too high to avoid repetitive transcription, which can happen sometimes. If it's being triggered unnecessarily regardless of the threshold we set, then that's an issue.
